### ISSUE TITLE
Memory leak using jboss jmx-remoting

### ISSUE THREAD

**Issue by adrianlzt** [rhuss/jolokia#401]
We are using jolokia in proxy mode. It is deployed like another war to a JBoss EAP 6.4 in domain mode.
To be able to connect to the JBoss JMX we have added the jboss-cli-client.jar to the jolokia.war.

After some days running, we see in the logs errors because OutOfMemory in the heap:
```
09:10:13,370 ERROR [org.xnio.listener] (Remoting "endpoint" read-1) A channel event listener threw an exception: java.lang.OutOfMemoryError: Java heap space
```

Making a memory heap dump and using Eclipse MAT to analyze it, it shows a clear responsible:
![image](https://user-images.githubusercontent.com/3237784/55106552-8e751180-50cf-11e9-8f43-cc66cffcfd2d.png)

Searching in the web I found [this issue](https://issues.jboss.org/browse/WFLY-820) (maybe not related) that points to some memory leaks problems with the file [JBossThreadFactory](https://github.com/jbossas/jboss-threads/blob/master/src/main/java/org/jboss/threads/JBossThreadFactory.java).
In that file the [last commit](https://github.com/jbossas/jboss-threads/commit/60872aa81b01d7733da9e1b64fa8a9b3fd1b1ddc) is ``[JBTHR-51] Proactively prevent TCCL leaks from poorly behaved tasks and thread factories``.
Which links to [this other issue](https://issues.jboss.org/browse/JBTHR-51) and that one to [this one](https://issues.jboss.org/browse/WFLY-9742).

In the last one, last comment says: ``WF 12 includes JBoss Threads 2.3.1, which has the related fix in it, so I'm marking this resolved in 12.``.
So I took the jboss-cli-client.jar of JBoss EAP 7.2 and copied the jboss-threads 2.3.1 to the jboss-cli-client.jar of the JBoss EAP 6.4 version (jar surgery).

But still having OutOfMemory problems in several jboss servers.

I was wondering if maybe Jolokia is not loading properly the jmx client and this could be causing this issue.
My java knowledge is quite basic, so it's hard to keep investigating.

Thanks!

**Comment by rhuss** [comment]
That sounds like a complex issue with many moving parts.

My first question is: Why do you need jboss-cli-client.jar for connecting to other JBoss servers via remote JMX ? The remote access is supposed to be standardized via JSR-160 so that no proprietary client libraries are supposed to be needed if you use the supported default RMI protocol used by a JSR-160 connector. You can add own protocols to JSR-160 connectors, then you need a client library. But I'm not sure if jboss-cli-client does this really.

Ideally, and to trim down the issue it could be a goal to get rid of this client library.

Otherwise, it's really hard to troubleshoot from a Jolokia perspective, especially as Jolokia itself does not start any threads (and the agent has never exposed a memory leak on its own in other setups).

As mentions in the docs, I consider the proxy mode only as the last choice if no other setup is possible. Beside potential performance issues, it also suffers from security issues. Most of them are fixed but because of the vulnerability of RMI, it still has a quite large attack surface. Maybe this could be an argument to switch to real agents, where Jolokia is installed directly on the target systems?

**Comment by adrianlzt** [comment]
Looks like the dropped RMI in JBoss EAP 6 (JBoss AS 7).
From a [RedHat solution](https://access.redhat.com/solutions/308643)
```
JBoss EAP 6 does not use RMI, so RMI classloading doesn't work because there is no RMI
```

Also from [this blog](https://blog.akquinet.de/2012/11/01/connecting-visualvm-with-a-remote-jboss-as-7-eap6-jvm-process/):
```
The application server supports the JMX Remote API for remote management and monitoring (JSR 160). Therefor, JBoss implements a JSR 160 compliant JMXConnector instead of an RMI interface to establish a connection to a JMX MBeanServer form a remote client
```

Our reason to use the proxy mode was that our jboss hosts run around 50 servers each one. Installing jolokia in each one could be too much overhead and some error in jolokia could impact the apps.

Thanks for the fast response BTW :)

**Comment by adrianlzt** [comment]
We opened a bug to RedHat and this was their response:

> It seems a possible workaround to mitigate this issue would be by reusing a JMXConnector object after it's obtained via JMXConnectorFactory.connect() in your JMX client application. 
> ...
> This only occurs when a large number of JMXConnector objects are created.
> 
> JMXConnector is an expensive object and should be created once and re-used.  It should not be created repeatedly for every call.
> This may indicate a problem in the Jolokia code.

I guess it is related with [createConnector]:(https://github.com/rhuss/jolokia/blob/8f0704beb25a6cdcf7fa91312e3d46252b8a0797/agent/jsr160/src/main/java/org/jolokia/jsr160/Jsr160RequestDispatcher.java#L116)
```
    // TODO: Add connector to a pool and release it on demand. For now, simply close it.
    private JMXConnector createConnector(JmxRequest pJmxReq) throws IOException {
```

**Comment by rhuss** [comment]
Yes, pooling is unfortunately not implemented yet ;-( (tbh, the proxy is not considered to be the first choice, and unfortunately I have not much time anymore to spend on Jolokia these days.

Happy though to help everyone who wants to jump and develop that missing feature.
